[Min Rei] iP

minreiseah opened the pull request:
# Derek 🦫

> Productivity.

Free your mind from distractions with this simple-to-use todolist.

**Features**:
1. Create tasks with `todo`, `event`, or `deadline`
2. List tasks with `list`
3. Mark (unmark) tasks as completed (uncompleted) with `mark` and `unmark`
4. Delete tasks with `delete`
5. Find tasks by keyword matcing with `find`

**Installation**:
- Download the jar file [here](https://github.com/minreiseah/ip/releases)
- Run it in your terminal of choice:
```bash
java -jar ip.jar
```
- Be productive

**Upcoming Features**:
- [x] Search functionality
- [ ] GUI
- [ ] Full Code Coverage

We are working on improving the code quality of `UserInterface.java` to reduce repeated code. The class' methods currently look like:

```java
public static void printTaskAdded(String msg, int numTasks) {
    System.out.println(INDENT + LINE);
    System.out.println(INDENT + "Added Task:");
    System.out.println(INDENT + "  " + msg);
    System.out.println(INDENT + "Now you have " + numTasks + " tasks in the list.");
    System.out.println(INDENT + LINE + "\n");
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

minreiseah committed:
Update gitignore

minreiseah committed:
Rename chatbot to Derek

minreiseah committed:
Echo user input

minreiseah committed:
Feat: Add, List tasks

minreiseah committed:
Feat: Mark tasks as done/undone

minreiseah committed:
Chore: Clean message output to console

minreiseah committed:
Feat: Add ToDo, Deadline, Event task types

minreiseah committed:
Refactor: user interface, command parser, exception, enums

minreiseah committed:
Test: Add automated testing

minreiseah committed:
Feat: Add Delete Task

minreiseah committed:
Add persistent task file storage

- TaskList was not persisted before
- TaskList and Task now implement Serializable
- Tasks serialized to tasklist.dat

minreiseah committed:
Merge branch 'branch-level-7' for Task Persistence

mahadhir247 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,128 @@
+                CommandType commandType = CommandParser.parseCommand(userInput);
+
+                switch (commandType) {
+                    case LIST:
```
Should the indentation of the case clauses follow the indentation of the switch statement?

mahadhir247 commented on src/main/java/PersistentStorageHandler.java:
```diff
@@ -0,0 +1,61 @@
+        return true;
+    }
+
+    private TaskList decode(ObjectInputStream ois) throws DukeException {
```
Perhaps this method name could be more descriptive? Such as decodeFileContents()?

mahadhir247 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,128 @@
+                    handleEvent(userInput);
+                    break;
+
+                    case EXIT:
+                    isExit = true;
+                    break;
+
+                    case BYE:
+                    isExit = true;
+                    break;
```
Any reason why you included both exit and bye? It seems that both these commands perform the same function?

mahadhir247 commented on src/main/java/PersistentStorageHandler.java:
```diff
@@ -0,0 +1,61 @@
+
+    public PersistentStorageHandler() { }
+
+    public boolean ensureTaskFileExists() throws DukeException {
```
I like how explicit this method name is 👍

mahadhir247 reviewed (commented):
Overall, most of your code quality follows the naming guidelines stated. A couple of methods could be a little bit more elaborate.

minreiseah committed:
Feat: Add datetime parsing

minreiseah committed:
Test: Update datetime testing

minreiseah committed:
Merge branch 'branch-Level-8'

minreiseah committed:
Refactor: Add packages

minreiseah committed:
Merge branch 'branch-A-packages'

legionlegion commented on src/main/java/duke/command/CommandParser.java:
```diff
@@ -0,0 +1,65 @@
+        }
+    }
+
+    public static String parseToDo(String userInput) throws DukeException {
```
Maybe it will be good to create new clases for the parsing functions? (I know u will in the future) ❤️

legionlegion commented on src/main/java/duke/storage/PersistentStorageHandler.java:
```diff
@@ -0,0 +1,67 @@
+import duke.commons.exceptions.TaskDataNotFoundException;
+import duke.task.TaskList;
+
+public class PersistentStorageHandler {
```
I like how you factored out the file handling class at such an early stage.

legionlegion commented on src/main/java/duke/Main.java:
```diff
@@ -0,0 +1,140 @@
+                String userInput = UserInterface.getUserInput();
+                CommandType commandType = CommandParser.parseCommand(userInput);
+
+                switch (commandType) {
```
Please change the indentation of switch and case statements to follow the standards of this mod! 🗿

legionlegion reviewed (commented):
LGTM! Just some nits to fix!

Hibeom0929 commented on src/main/java/duke/Main.java:
```diff
@@ -0,0 +1,140 @@
+
+                    case BYE:
+                        isExit = true;
+                        break;
```
Note that there is no indentation for case clauses! The indentation of the switch and case should be the same.

Hibeom0929 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,29 @@
+
+public abstract class Task implements Serializable {
+    private String description;
+    private boolean isDone;
```
I like how you name the boolean.

Hibeom0929 commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,71 @@
+            System.out.println(INDENT + "no tasks");
+        } else {
+            for (int i = 0; i < tasks.size(); i++) {
+                System.out.println("     "  + (i+1) + ". " + tasks.get(i));
```
I think it would be better to make operators surrounded by a space character.
```suggestion
                System.out.println("     "  + (i + 1) + ". " + tasks.get(i));
```

Hibeom0929 reviewed (commented):
LGTM! Your code is clean and easy to understand. I noted a few minor coding standard violations too.

minreiseah committed:
Update gitignore

minreiseah committed:
Merge branch 'branch-A-Jar'

minreiseah committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle
Add Gradle Support

minreiseah committed:
Update gitignore

minreiseah committed:
Merge branch 'branch-A-Gradle'

minreiseah committed:
Update gradle build main class

minreiseah committed:
Set gradle build to java 11

minreiseah committed:
Add command parser tests

minreiseah committed:
Add JavaDocs

minreiseah committed:
Fix minor breaches to coding standard

minreiseah committed:
Add find functionality to search for tasks containing keywords

The user should be able to search their tasks, especially in cases where
there are many tasks.

Keyword matching was implemented to allow the user to search with
multiple keywords across all their tasks.

minreiseah committed:
Add support for more datetime formats.

Event and Deadline can only accept the default format of java's
DateTiemFormatter.

Add a new DateUtils class to iterate over a wider selection of formats.

More robust datetime reader allows user to input more datetime formats.

minreiseah committed:
Merge branch 'branch-A-CodingStandard'

minreiseah committed:
Merge branch 'branch-A-JavaDoc'

minreiseah committed:
Update README

minreiseah committed:
Merge branch 'branch-Level-9'

minreiseah committed:
Add checkstyle

minreiseah committed:
Merge branch 'branch-A-Checkstyle'

minreiseah committed:
Update Gradle

- Disabled Checkstyle on build
- Added JavaFX Dependencies

minreiseah committed:
Add basic echo GUI

minreiseah committed:
Upgraded fully to GUI

minreiseah committed:
Merge branch 'branch-Level-10'

minreiseah committed:
Update Gradle Build

minreiseah committed:
Add basic assertions

minreiseah committed:
Refactor out controller

minreiseah committed:
Refactor out UI

minreiseah committed:
Merge pull request #2 from minreiseah/branch-A-Assertions

A-Assertions

minreiseah committed:
Merge branch 'master' of https://github.com/minreiseah/ip

minreiseah committed:
Merge branch 'master' into branch-A-CodeQuality

minreiseah committed:
Merge pull request #3 from minreiseah/branch-A-CodeQuality

A-CodeQuality

minreiseah committed:
Merge branch 'master' of https://github.com/minreiseah/ip

minreiseah committed:
Add task duplication detection

minreiseah committed:
Add task testing

minreiseah committed:
Merge branch 'branch-C-DetectDuplicates'

minreiseah committed:
Fix tasklist not saving

minreiseah committed:
Add human-readable file storage

minreiseah committed:
Update user guide
